### ISSUE TITLE
1주차 suhshin : emotion과 styled-component의 차이

### ISSUE THREAD

**Issue by rkskekzzz** [rkskekzzz/apdan#1]
채용 공고 페이지를 보면 emotion을 사용하는 부서가 많습니다!

차이는 어떤지, 왜 emotion을 선택하는지 기존의 styled-component로 된 코드를 직접 리팩토링을 진행하면서 알아보려고 합니다!

* [깃허브 링크](https://github.com/rkskekzzz/react-playground/tree/main/emotion_vs_styledcomponents)
* [블로그 링크](https://80000coding.oopy.io/7ad296c7-8832-4951-9cf7-074a196d42ea)

**Comment by mike2ox** [comment]
styel-component만 썼어서 몰랐는데 emotion의 스타일 적용방식을 알아볼 수 있어서 좋았습니다 👍 
블로그 내용과 예제 코드를 돌려보는 중에 생각난 부분 공유할게요!

1. 패키지 다운로드 수를 비교하시던데 emotion/core와 emotion/react 사용 유저간의 독립성이 보장되지 않아 2배보단 비슷한 비율로 사용중으로 보는게 제공 자료 해석의 최선일 것 같아요

2. ~emotion 사용시 주의사행에 ts에 적용할 경우 pragma line(**/**@jsxImportSource @emotion/react */**)을 안써도 되는건지 아니면 pragma line도 써주고 comfileOptions도 적어줘야 할까요? 제공해주신 코드는 js기반이라서 질문 드립니다.~
 -> codesandbox만 js고 레포에는 ts라서 레포 코드로 질문한 부분 확인했습니다! (둘다 써줘야 하네요)

3. 태그의 클래스네임으로 디버깅하는 방법은 따로 생각안해봤는데 styled-component이든 emotion이든 구조가 큰 사이트의 스타일 디버깅할 때 도움이 많이 될거 같네요. 감사합니다 :)

4. 제공하신 @emotion/react로 작성된 코드에서 emotionComponent의 버튼들을 클릭하니 하단의 사진처럼 스타일 태그들이 쌓이는데 이게 일반적인 emotion의 동작인지 궁금합니다!
![image](https://user-images.githubusercontent.com/22931103/188689492-01795b96-e597-40d3-bcbf-238269578710.png)

**Comment by 6mn12j** [comment]
작성해주신 내용 잘 봤습니다!! emotion을 사용하는 곳들이 많아서 궁금 했는데 정리해주신 내용 보니까 스타일드 컴포넌트랑 비슷한 느낌인거 같네요

- 코드 딱 봤을때 가장 큰 차이점은 이모션은 임포트해서 css속성에 넣는거 같습니다.
- 이모션 css안에 ${button} 요부분은 어떤 값인가요??
- 디버깅 시 클래스이름이 항상 변환된 부분이 있어서 보기 불편 했는데 ‘label’ 기능은 알고 있으면 다음에 디버깅 할때 매우 편할거 같아요 👍👍
수고하셨습니다!

**Comment by nyeoni** [comment]
너무 깔끔하게 정리해주셔서 많은 도움이 되었습니다 감사합니다 :) `emotion` 을 많이 쓰는 이유가 번들사이즈가 작아서 가볍기 때문일거 같네요!! ㅎㅎ 

**질문**
1.  전에 `emotion` 을 썼을때는 `styled-component` 와 비슷하게 `styled.<tag>` 로 사용했었는데 `css` 방식으로 사용하셨더라구요!! 혹시 이 두 방법의 차이점이 있을까요?? 
2. `/** @jsxImportSource @emotion/react */` 를 `emotion` 쓸때 추가해줘야한다고 적혀있는데 그 이유를 알고 싶습니다! 제가 사용했었을때는 안 추가했었던 거 같은데 어떤 경우에 저 태그를 추가해줘야 하는지 궁금합니다!

**Comment by rkskekzzz** [comment]
@mike2ox 

> 패키지 다운로드 수를 비교하시던데 emotion/core와 emotion/react 사용 유저간의 독립성이 보장되지 않아 2배보단 비슷한 비율로 사용중으로 보는게 제공 자료 해석의 최선일 것 같아요

맞아요 독립성이 보장되진 않지만, core와 react는 rename된 패키지이기 때문에 두개가 동시에 설치되는 경우는 거의 없다고 보는게 맞을 것 같아요! 기존 core를 사용하던 서비스는 core를 사용하고 새로운 서비스들이 react패키지를 사용하는 것 같습니다~̆̈!
<img width="346" alt="image" src="https://user-images.githubusercontent.com/57925497/188846277-a81d8934-f66a-4951-b8c3-6871c36e86ce.png">

> 제공하신 @emotion/react로 작성된 코드에서 emotionComponent의 버튼들을 클릭하니 하단의 사진처럼 스타일 태그들이 쌓이는데 이게 일반적인 emotion의 동작인지 궁금합니다!

개발버전에서 동작하는 방식으로 알고있습니당

**Comment by rkskekzzz** [comment]
@6mn12j 
> 이모션 css안에 ${button} 요부분은 어떤 값인가요??

코드를 보면 button이라는 동일한 속성을 bgButton, fgButton에 적용시키기 위한 방법입니다~̆̈!
```typescript
const button = css`
  border: 0px solid black;
  font-size: 20px;
  border-radius: 8px;
  box-shadow: -2px 2px 4px rgba(0, 0, 0, 0.6);
  cursor: pointer;
`;

const bgButton = css`
  ${button}
  background-color: white;
  color: black;
  &:hover {
    background-color: black;
    color: white;
  }
`;

const fgButton = css`
  ${button}
  background-color: black;
  color: white;
  &:hover {
    background-color: white;
    color: black;
  }
`;
```

**Comment by rkskekzzz** [comment]
@Chloekkk 

> 전에 emotion 을 썼을때는 styled-component 와 비슷하게 styled.<tag> 로 사용했었는데 css 방식으로 사용하셨더라구요!! 혹시 이 두 방법의 차이점이 있을까요??

차이점은 크게 없습니다! 방식의 차이라고 볼 수 있어요! @emotion/styled api를 활용하면 동일하게 사용할 수 있습니다! 컴포넌트에서 css prop을 넘겨주어야 하는 경우에는 styled 방식이 조금 더 편해 보였습니다 ㅎㅎ

> /** @jsxImportSource @emotion/react */ 를 emotion 쓸때 추가해줘야한다고 적혀있는데 그 이유를 알고 싶습니다! 제가 사용했었을때는 안 추가했었던 거 같은데 어떤 경우에 저 태그를 추가해줘야 하는지 궁금합니다!

@emotion/styled api를 활용하면 문제 없지만 @emotion/react 속성을 사용하면 jsx component에 css라는 속성이 react패키지에서는 가지고 있지 않기 때문에 babel에게 emotion의 jsx를 사용해서 트랜스파일링 하라고 알려주는 것입니다~̆̈!
